consumers/writers/timescale/consumer.go

```diff
@@ -0,0 +1,203 @@
+          name, unit, value, string_value, bool_value, data_value, sum,
+          time, update_time)
+          VALUES (:channel, :subtopic, :publisher, :protocol, :name, :unit,
+          :value, :string_value, :bool_value, :data_value, :sum,
+          :time, :update_time);`
+
+	tx, err := pr.db.BeginTxx(context.Background(), nil)
+	if err != nil {
+		return errors.Wrap(errSaveMessage, err)
+	}
+	defer func() {
+		if err != nil {
+			if txErr := tx.Rollback(); txErr != nil {
+				err = errors.Wrap(err, errors.Wrap(errTransRollback, txErr))
+			}
+			return
+		}
+
+		if err = tx.Commit(); err != nil {
+			err = errors.Wrap(errSaveMessage, err)
+		}
+	}()
+
+	for _, msg := range msgs {
+		m := senmlMessage{Message: msg}
+		if _, err := tx.NamedExec(q, m); err != nil {
+			pqErr, ok := err.(*pq.Error)
+			if ok {
+				switch pqErr.Code.Name() {
+				case errInvalid:
+					return errors.Wrap(errSaveMessage, errInvalidMessage)
+				}
+			}
+
+			return errors.Wrap(errSaveMessage, err)
+		}
+	}
+	return err
+}
+
+func (pr timescaleRepo) saveJSON(msgs mfjson.Messages) error {
+	if err := pr.insertJSON(msgs); err != nil {
+		if err == errNoTable {
+			if err := pr.createTable(msgs.Format); err != nil {
+				return err
+			}
+			return pr.insertJSON(msgs)
+		}
+		return err
+	}
+	return nil
+}
+
+func (pr timescaleRepo) insertJSON(msgs mfjson.Messages) error {
+	tx, err := pr.db.BeginTxx(context.Background(), nil)
+	if err != nil {
+		return errors.Wrap(errSaveMessage, err)
+	}
+	defer func() {
+		if err != nil {
+			if txErr := tx.Rollback(); txErr != nil {
+				err = errors.Wrap(err, errors.Wrap(errTransRollback, txErr))
+			}
+			return
+		}
+
+		if err = tx.Commit(); err != nil {
+			err = errors.Wrap(errSaveMessage, err)
+		}
+	}()
+
+	q := `INSERT INTO %s (channel, created, subtopic, publisher, protocol, payload)
+          VALUES (:channel, :created, :subtopic, :publisher, :protocol, :payload);`
+	q = fmt.Sprintf(q, msgs.Format)
+
+	for _, m := range msgs.Data {
+		var dbmsg jsonMessage
+		dbmsg, err = toJSONMessage(m)
+		if err != nil {
+			return errors.Wrap(errSaveMessage, err)
+		}
+		if _, err = tx.NamedExec(q, dbmsg); err != nil {
+			pqErr, ok := err.(*pq.Error)
+			if ok {
+				switch pqErr.Code.Name() {
+				case errInvalid:
+					return errors.Wrap(errSaveMessage, errInvalidMessage)
+				case errUndefinedTable:
+					return errNoTable
+				}
+			}
+			return err
+		}
+	}
+	return nil
+}
+
+func (pr timescaleRepo) createTable(name string) error {
+	q := `CREATE TABLE IF NOT EXISTS %s (
+                        created       BIGINT NOT NULL,
+                        channel       VARCHAR(254),
+                        subtopic      VARCHAR(254),
+                        publisher     VARCHAR(254),
+                        protocol      TEXT,
+                        payload       JSONB,
+                        PRIMARY KEY (created, publisher, subtopic)
+                    );`
```

Comment:
It's ok, it's just that JSON doesn't really use any of the benefits of timescaledb - high cardinality indexes, continuous aggregates.  I've left the JSON option in there for completeness.

Reply:
Ah ok..  I re-used the indentation from the existing postgres consumer but happy to change if you need.

Reply:
In that case maybe you can fix it in both places?

Reply:
Ok.  Could you have a look and see how many spaces there should be at the beginning of each line?
It's hard to tell as it will look different compared to tabs based on the editor you look at it in.

Reply:
TBH I'm never really sure with this multines blocks but I'd say something liek that:

```
q := `CREATE TABLE IF NOT EXISTS %s (
        created       BIGINT NOT NULL,
        channel       VARCHAR(254),
        subtopic      VARCHAR(254),
        publisher     VARCHAR(254),
        protocol      TEXT,
        payload       JSONB,
        PRIMARY KEY (created, publisher, subtopic)
        );`
```

Reply:
ha, yes me too, ok, i've edited the indentation, and renamed repo -> tr